server/src/main/java/io/crate/execution/ddl/tables/AddColumnTask.java

```diff
@@ -123,34 +112,56 @@ public ClusterState execute(ClusterState currentState, AddColumnRequest request)
+
+    /**
+     * Adds missing parent columns.
+     * Also tries to update request by taking existing references from the cluster state.
+     * @return NULL if all columns already exist.
+     */
+    @Nullable
+    static AddColumnRequest normalizeColumns(AddColumnRequest request, DocTableInfo currentTable) {
         List<Reference> newColumns = new ArrayList<>();
-        boolean anyMissing = false;
-        for (var newColumn : request.references()) {
-            if (newColumn.column().isTopLevel()) {
-                newColumns.add(newColumn);
+
+        // Get existing references
+        boolean allExist = true;
+        for (Reference ref : request.references()) {
+            Reference exists = currentTable.getReference(ref.column());
+            if (exists == null) {
+                allExist = false;
+                newColumns.add(ref); // Actually new column, take it from request.
             } else {
-                ColumnIdent parent = newColumn.column();
-                while ((parent = parent.getParent()) != null) {
-                    if (!Symbols.containsColumn(request.references(), parent)
-                            && !Symbols.containsColumn(newColumns, parent)) {
-                        newColumns.add(currentTable.getReference(parent));
-                        anyMissing = true;
-                    }
+                if (ref.valueType().id() != exists.valueType().id()) {
+                    throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                        "Column `%s` already exists with type `%s`. Cannot add same column with type `%s`",
+                        ref.column(),
+                        exists.valueType().getName(),
+                        ref.valueType().getName()));
                 }
-                newColumns.add(newColumn);
+                newColumns.add(exists); // Get updated reference from the cluster state.
             }
         }
-        if (anyMissing) {
-            return new AddColumnRequest(
-                request.relationName(),
-                newColumns,
-                request.checkConstraints(),
-                request.pKeyIndices()
-            );
-        } else {
-            return request;
+        if (allExist) {
+            return null;
         }
+
+        // Add missing parents
+        for (var newColumn : request.references()) {
+            ColumnIdent parent = newColumn.column();
+            while ((parent = parent.getParent()) != null) {
+                if (!Symbols.containsColumn(request.references(), parent)
+                        && !Symbols.containsColumn(newColumns, parent)) {
+                    newColumns.add(currentTable.getReference(parent));
+                }
+            }
+        }
+
+        // If allExist was not short-circuited, we have at least one updated reference, have to copy request regardless of missing parents presence.
```

Comment:
It would also be possible to change this to only create a new `List<Reference>`. Only the columns change. Contraints, pkIndices and relation name remain the same.